c10/util/Optional.h

```diff
@@ -964,7 +964,7 @@ constexpr bool operator==(const optional<T>& x, const optional<T>& y) {
 
 template <class T>
 constexpr bool operator!=(const optional<T>& x, const optional<T>& y) {
-  return !(x == y);
+  return x ? *x != y : bool(y);
```

Comment:
Why isn't the implementation here simply:

```
return bool(x) != bool(y) ? true : bool(x) == false ? false : *x != *y;
```

in symmetry with the implementation above?

Reply:
The implementation above is not simpler and appears to be written by someone who doesn't really understand booleans.

It's almost always a code smell to use the equality operator on a boolean literal (e.g. `bool(x) == false`). And using a ternary operator to return boolean literals also makes little sense.

The nested ternary also makes the logic a bit hard to follow, IMO.

IMO we could simplify all the implementations, but I don't think that's a good use of time since we're fairly close to C++17.

This PR is mostly to address a subtle correctness issue where we diverge from the spec.

Reply:
Writing standard library code in C++ is very subtle, so having good hygiene when making modifications like this is very important, in my opinion. If the modification you made is not equivalent to the old code, you will be in a horrible situation where `==` and `!=` have different semantics just because the code is written differently and you didn't update the old call sites. Better hygiene is if you opened a separate PR to switch all of the statements to the new style, and then made the bugfix here. (BTW, we're not *that* close to C++17; per https://github.com/pytorch/pytorch/issues/56055 we have to wait for CUDA 10 to die and it seems to be dying... slowly.)

Now I'm going to show you that the new implementation is not semantically equivalent to the old one. I extracted your implementation and the old implementation into this program:

```
#include <optional>

bool neq1(void* v, std::optional<void*> x) {
    return bool(x) ? v != *x : true;
}

extern "C" bool tgt(std::optional<void*> x, std::optional<void*> y) {
    return x ? neq1(*x, y) : bool(y);
}

extern "C" bool src(std::optional<void*> x, std::optional<void*> y) {
    return !(bool(x) != bool(y) ? false : bool(x) == false ? true : *x == *y);
}
```

Using clang 12.0.1 and flags `-O2 -g0 -std=c++17 -emit-llvm` (https://godbolt.org/z/Yqo7n7398) I extracted the LLVM IR and then compared src and tgt using Alive2, a translation validation tool (https://alive2.llvm.org/ce/z/g-T9Zz) to see if it is valid to replace the original (src) with your new implementation (tgt). It is not.

```
ERROR: Target is more poisonous than source

Example:
* %0 = poison
i8 %1 = #xff (255, -1)
* %2 = poison
i8 %3 = #x00 (0)

...

Source value: #x1 (1)
Target value: poison
```

Specifically, when the "has value" field of both input optionals is poisoned (i.e. undefined behavior), but  the values in the optionals are not, the original implementation is able to report that the optionals are definitely not equal. In your implementation, the result in this case is also undefined behavior (because you force a ternary on these values).

I also extracted GCC's implementation of optional (which doesn't involve ternaries and I do agree is much more pleasing than the old implementation) and showed that it is equivalent to the old code but not the new code.

```
extern "C" bool src(std::optional<void*> x, std::optional<void*> y) {
    return bool(x) != bool(y) || (bool(x) && *x != *y);
}
```

So this modification is just not right, and we shouldn't accept it.

Reply:
@swolchok mentions to me that it's pretty unlikely for has_value to become poisoned in practice, because the constructor should initialize it, and moving from standard types should leave it unspecified but valid.

Reply:
The goal of this is not to have identical behavior to the previous implementation, though I did not intend to change it in this particular way. Very interesting!

I agree that it makes sense to make our implementations consistent, minimally from a style perspective. I don't believe that the priority of doing that is high though, **but it's easy enough and I'm willing to do it**. It would be less work than this discussion, but much less fun :).

However, unless I'm misunderstanding you, this implementation is not incorrect. It is well defined and correct for valid inputs. If the inputs are not valid, the output is also not valid. But that's ok, because the program is invalid since we already produced undefined behavior prior to entering this function.

IMO, it doesn't matter if an invalid input produces an invalid output. In fact I think it's counterproductive to consider invalid inputs when defining functions. Where does that line of thinking end?

I **do** think that there's merit in having the exact same implementation as a quality implementation of `std::optional` (libstdc++ or libc++) since we do eventually want to switch to those. **So I'm not opposed to replicating that (and that I'm willing to let trump anything).**

Reply:
> However, unless I'm misunderstanding you, this implementation is not incorrect. It is well defined and correct for valid inputs. If the inputs are not valid, the output is also not valid. But that's ok, because the program is invalid since we already produced undefined behavior prior to entering this function.

This is a misunderstanding of how UB works in LLVM. There are certainly classes of undefined behavior where, if it happens, all bets are off, the nuclear missiles can be fired. But there are also cases where there is undefined behavior, but not *that* undefined, and undef/poison are examples of this. undef you can think of as "reading from uninitialized memory", poison you can think of as "you overflowed an integer".  The Alive2 paper has a pretty succinct discussion of how these are handled inside LLVM https://www.cs.utah.edu/~regehr/alive2-pldi21.pdf

Reply:
Undefined behavior is a runtime concept: the text of a program doesn't have undefined behavior, only the execution of a program with some inputs is undefined behavior. (This is why the program `x+1` isn't UB, even though x could be MAX_INT and overflow). But the life of a compiler is all about moving instructions around and generally changing the source program, and so a compiler might want to, for example, hoist `x+1` out of a loop and "Yes, OK, maybe that could have caused UB, but please don't blow up immediately because in the source program this only would have happened in the loop, so please blow up *later* if you actually would have ended up in the loop", which is where the concepts of poison and undef come from.

To put it differently, reasoning about the behavior of a function under undef/poison helps us reason about how the program interacts with other code that might have UB (but then again, might not! But the compiler doesn't know what input it's actually going to run on, it's got to optimize compatibly with all possible inputs.)